In [1]:
# Initial state (S0): empty 3x3 board
S0 = [" " for _ in range(9)]

def print_board(state):
    print(state[0]+"|"+state[1]+"|"+state[2])
    print("-+-+-")
    print(state[3]+"|"+state[4]+"|"+state[5])
    print("-+-+-")
    print(state[6]+"|"+state[7]+"|"+state[8])


def Player(state):
    """Return which player has the next turn"""
    X_count = state.count("X")
    O_count = state.count("O")
    return "X" if X_count == O_count else "O"


def Actions(state):
    """Return all legal moves"""
    return [i for i in range(9) if state[i] == " "]


def Result(state, action):
    """Return new state after action is taken"""
    player = Player(state)
    new_state = state.copy()
    new_state[action] = player
    return new_state


def Terminal(state):
    """True if someone wins or board is full"""
    win_states = [
        [0,1,2], [3,4,5], [6,7,8],      # rows
        [0,3,6], [1,4,7], [2,5,8],      # columns
        [0,4,8], [2,4,6]                # diagonals
    ]
    # Check win
    for a,b,c in win_states:
        if state[a] != " " and state[a] == state[b] == state[c]:
            return True

    # Check draw
    return " " not in state


def Utility(state):
    """Return +1 if X wins, -1 if O wins, else 0"""
    win_states = [
        [0,1,2], [3,4,5], [6,7,8],
        [0,3,6], [1,4,7], [2,5,8],
        [0,4,8], [2,4,6]
    ]
    for a,b,c in win_states:
        if state[a] == state[b] == state[c] and state[a] != " ":
            return 1 if state[a] == "X" else -1
    return 0


In [2]:
call_count = 0  # global counter

def MAX_VALUE(state):
    global call_count
    call_count += 1

    if Terminal(state):
        return Utility(state)

    v = float("-inf")
    for action in Actions(state):
        v = max(v, MIN_VALUE(Result(state, action)))
    return v


def MIN_VALUE(state):
    global call_count
    call_count += 1

    if Terminal(state):
        return Utility(state)

    v = float("inf")
    for action in Actions(state):
        v = min(v, MAX_VALUE(Result(state, action)))
    return v


def minimax_decision(state):
    """Return best move for current player"""
    global call_count
    call_count = 0  # reset per search

    player = Player(state)
    best_move = None

    if player == "X":
        best_val = float("-inf")
        for action in Actions(state):
            val = MIN_VALUE(Result(state, action))
            if val > best_val:
                best_val = val
                best_move = action
    else:
        best_val = float("inf")
        for action in Actions(state):
            val = MAX_VALUE(Result(state, action))
            if val < best_val:
                best_val = val
                best_move = action

    print("States explored =", call_count)
    return best_move


In [3]:
def play():
    state = S0.copy()

    while not Terminal(state):
        print_board(state)

        if Player(state) == "X":      # Human
            move = int(input("Enter your move (1-9): ")) - 1
            if move not in Actions(state):
                print("Invalid move! Try again.")
                continue
        else:                         # AI
            print("AI is thinking...")
            move = minimax_decision(state)

        state = Result(state, move)

    print_board(state)
    if Utility(state) == 1:
        print("X wins!")
    elif Utility(state) == -1:
        print("O wins!")
    else:
        print("It's a draw!")


In [6]:
play()

 | | 
-+-+-
 | | 
-+-+-
 | | 
Enter your move (1-9): 5
 | | 
-+-+-
 |X| 
-+-+-
 | | 
AI is thinking...
States explored = 55504
O| | 
-+-+-
 |X| 
-+-+-
 | | 
Enter your move (1-9): 9
O| | 
-+-+-
 |X| 
-+-+-
 | |X
AI is thinking...
States explored = 1172
O| |O
-+-+-
 |X| 
-+-+-
 | |X
Enter your move (1-9): 2
O|X|O
-+-+-
 |X| 
-+-+-
 | |X
AI is thinking...
States explored = 50
O|X|O
-+-+-
 |X| 
-+-+-
 |O|X
Enter your move (1-9): 4
O|X|O
-+-+-
X|X| 
-+-+-
 |O|X
AI is thinking...
States explored = 4
O|X|O
-+-+-
X|X|O
-+-+-
 |O|X
Enter your move (1-9): 7
O|X|O
-+-+-
X|X|O
-+-+-
X|O|X
It's a draw!


# **PART D:**

## ***Question***: W*hy is the actual state count (~55,504) smaller than the theoretical 109,600?*

## ***Ans***:
Because many games end early, so Minimax stops exploring deeper moves.
Symmetry reduces unique states, and invalid/after-win sequences are not explored.
Therefore, the real tree is much smaller than the full theoretical one.

# **PART E:**

**1. Why does Minimax evaluate many states?**

1. It must have to consider all possible future moves to guarantee optimal play.

**2. Full tree vs. actual explored states?**

2. Full tree includes every possible sequence; Minimax explores only legal or actual, needed states.

**3. How do early wins reduce states?**

3. A win cuts off the branch, so deeper states are never explored.

**4. Challenges applying Minimax to Chess?**

4. Huge state space; requires pruning, heuristics, depth limits, and optimizations.